pilot/pkg/config/kube/gateway/inferencepool_collection.go

```diff
@@ -0,0 +1,317 @@
+				LastTransitionTime: metav1.NewTime(time.Now()),
+			},
+		},
+	}
+}
+
+// generateHash generates an 8-character SHA256 hash of the input string.
+func generateHash(input string, length int) string {
+	hashBytes := sha256.Sum256([]byte(input))
+	hashString := fmt.Sprintf("%x", hashBytes) // Convert to hexadecimal string
+	return hashString[:length]                 // Truncate to desired length
+}
+
+func InferencePoolServiceName(poolName string) (string, error) {
+	ipSeparator := "-ip-"
+	hash := generateHash(poolName, hashSize)
+	svcName := poolName + ipSeparator + hash
+	// Truncate if necessary to meet the Kubernetes naming constraints
+	if len(svcName) > maxServiceNameLength {
+		// Calculate the maximum allowed base name length
+		maxBaseLength := maxServiceNameLength - len(ipSeparator) - hashSize
+		if maxBaseLength < 0 {
+			return "", fmt.Errorf("inference pool name: %s is too long", poolName)
+		}
+
+		// Truncate the base name and reconstruct the service name
+		truncatedBase := poolName[:maxBaseLength]
+		svcName = truncatedBase + ipSeparator + hash
+	}
+	return svcName, nil
+}
+
+// translateInferencePoolToService converts an InferencePool to a Service
+func translateInferencePoolToService(pool *inferencev1alpha2.InferencePool, extRef extRefInfo) (*corev1.Service, error) {
+	svcName, err := InferencePoolServiceName(pool.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	shadowSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: pool.GetNamespace(),
+			Labels: map[string]string{
+				InferencePoolRefLabel:              pool.GetName(),
+				InferencePoolExtensionRefSvc:       extRef.name,
+				InferencePoolExtensionRefPort:      strconv.Itoa(int(extRef.port)),
+				constants.InternalServiceSemantics: constants.ServiceSemanticsInferencePool,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector:  make(map[string]string, len(pool.Spec.Selector)),
+			Type:      corev1.ServiceTypeClusterIP,
+			ClusterIP: corev1.ClusterIPNone, // Headless service
+			Ports: []corev1.ServicePort{ // adding dummy port, not used for anything
+				{
+					Protocol:   "TCP",
+					Port:       int32(54321),
+					TargetPort: intstr.FromInt(int(pool.Spec.TargetPortNumber)),
+				},
+			},
+		},
+	}
+
+	for k, v := range pool.Spec.Selector {
+		shadowSvc.Spec.Selector[string(k)] = string(v)
+	}
+
+	shadowSvc.SetOwnerReferences([]metav1.OwnerReference{
+		{
+			APIVersion: gvk.InferencePool.GroupVersion(),
+			Kind:       gvk.InferencePool.Kind,
+			Name:       pool.GetName(),
+			UID:        pool.GetUID(),
+		},
+	})
+	return shadowSvc, nil
+}
+
+func (c *Controller) reconcileShadowService(
+	ctx krt.HandlerContext,
+	service *corev1.Service,
+	servicesCollection krt.Collection[*corev1.Service],
+) error {
+	// Get the existing service if it exists
+	key := service.Namespace + "/" + service.Name
+	// N.B we probably do want to subscribe to changes here e.g. so we reconcile if the
```

Comment:
wdym?

Reply:
Everytime you do krt.Fetch, it subscribes the collection you're building to changes from the collection you're fetching from. It may seem wasteful to do that here in the current implementation since we're creating a service (which is going to trigger another reconcile), but it is useful for making sure the user doesn't mess with these services